### ISSUE TITLE
OOM with loading ~500 images in gridview

### ISSUE THREAD

**Issue by jcelaviie** [bumptech/glide#842]
I have hit OOM issues every time I open sticker picker. The code looks like the following. 
I tried to clear the adapter and set imageView drawable in getView callback, but the memory usage never decrease.

Thanks a lot!

```java
public class MyFragment extends BinderFragment {
  
  GridView stickerGridView;
  String albumId;
  AccountHandler accountHandler;

  @Override
  protected void onAttachBinder(Bundle savedInstanceState) {
    super.onAttachBinder(savedInstanceState);
    accountHandler = binder.get(AccountHandler.class);
  }

  private class GalleryItem {
    public String photoId;
    public String photoUrl;
    public String photoCacheFile;
  }

  private class GalleryItemAdapter extends ArrayAdapter<GalleryItem> {

    public GalleryItemAdapter(ArrayList<GalleryItem> items) {
      super(getActivity(), 0, items);
    }

    @Override
    public View getView(final int position, View convertView, ViewGroup parent) {
      if (convertView == null) {
        convertView = getActivity().getLayoutInflater()
            .inflate(R.layout.gallery_item_v2, parent, false);
      }

      final GalleryItem item = getItem(position);
      ImageView imageView = (ImageView) convertView.findViewById(R.id.gallery_item_image);

     
      // use Glide to load image    
      Glide.with(context)
          .asDrawable()
          .apply(requestionOptions)
          .load(item.photoUrl)
          .into(imageView);

      return convertView;
    }

@Override
  public void onDestroy() {
    for(int i=0; i< stickerGridView.getChildCount(); i++) {
      View view = stickerGridView.getChildAt(i);
      Glide.with(context).clear(view);
    }
    super.onDestroy();
  }

    @Override
    protected void onPostExecute(ArrayList<GalleryItem> items) {
      setupAdapter(stickerGridView, items);
    }
  }

  @Override
  public void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    albumId = getArguments().getString(Intents.EXTRA_ALBUM_ID);
  }

  @Override
  public View onCreateView(LayoutInflater inflater, ViewGroup container,
      Bundle savedInstanceState) {
    stickerGridView =
        (GridView) inflater.inflate(R.layout.sticker_gallery_v2,
            container, false);
    setupAdapter(stickerGridView, null);

    stickerGridView.setOnItemClickListener(new OnItemClickListener() {
      @Override
      public void onItemClick(AdapterView<?> gridView, View view, int pos, long id) {
        sendSticker(pos);
      }
    });
    return stickerGridView;
  }

 
  void setupAdapter(GridView gridView, ArrayList<GalleryItem> items) {
    if (getActivity() == null || gridView == null) {
      return;
    }
    if (items != null) {
      gridView.setAdapter(new GalleryItemAdapter(items));
    } else {
      gridView.setAdapter(null);
    }
  }

  @Override
  public void onDestroyView() {
    LogUtil.i(TAG, "Clearing cache.");
    super.onDestroyView();
  }
}
```

```
E System  : java.lang.OutOfMemoryError: Failed to allocate a 24 byte allocation with 3832 free bytes and 3KB until OOM
12-25 14:24:24.437  5798  5806 E System  : 	at com.android.internal.os.BinderInternal$GcWatcher.finalize(BinderInternal.java:53)
12-25 14:24:24.437  5798  5806 E System  : 	at java.lang.Daemons$FinalizerDaemon.doFinalize(Daemons.java:202)
12-25 14:24:24.437  5798  5806 E System  : 	at java.lang.Daemons$FinalizerDaemon.run(Daemons.java:185)
12-25 14:24:24.437  5798  5806 E System  : 	at java.lang.Thread.run(Thread.java:818)
12-25 14:24:24.457  5798  7575 E GlideExecutor: Request threw uncaught throwable
12-25 14:24:24.457  5798  7575 E GlideExecutor: java.lang.OutOfMemoryError: Failed to allocate a 47 byte allocation with 520 free bytes and 520B until OOM
12-25 14:24:24.457  5798  7575 E GlideExecutor: 	at android.graphics.Bitmap.nativeCreate(Native Method)
12-25 14:24:24.457  5798  7575 E GlideExecutor: 	at android.graphics.Bitmap.createBitmap(Bitmap.java:831)
12-25 14:24:24.457  5798  7575 E GlideExecutor: 	at android.graphics.Bitmap.createBitmap(Bitmap.java:808)
12-25 14:24:24.457  5798  7575 E GlideExecutor: 	at android.graphics.Bitmap.createBitmap(Bitmap.java:775)
12-25 14:24:24.457  5798  7575 E GlideExecutor: 	at com.bumptech.glide.load.engine.bitmap_recycle.LruBitmapPool.get(LruBitmapPool.java:129)
12-25 14:24:24.457  5798  7575 E GlideExecutor: 	at com.bumptech.glide.load.resource.bitmap.TransformationUtils.fitCenter(TransformationUtils.java:140)
12-25 14:24:24.457  5798  7575 E GlideExecutor: 	at com.bumptech.glide.load.resource.bitmap.TransformationUtils.centerInside(TransformationUtils.java:181)
12-25 14:24:24.457  5798  7575 E GlideExecutor: 	at com.bumptech.glide.load.resource.bitmap.CenterInside.transform(CenterInside.java:33)
12-25 14:24:24.457  5798  7575 E GlideExecutor: 	at com.bumptech.glide.load.resource.bitmap.BitmapTransformation.transform(BitmapTransformation.java:76)
12-25 14:24:24.457  5798  7575 E GlideExecutor: 	at com.bumptech.glide.load.resource.gif.GifDrawableTransformation.transform(GifDrawableTransformation.java:45)
12-25 14:24:24.457  5798  7575 E GlideExecutor: 	at com.bumptech.glide.load.engine.DecodeJob$DecodeCallback.onResourceDecoded(DecodeJob.java:474)
12-25 14:24:24.457  5798  7575 E GlideExecutor: 	at com.bumptech.glide.load.engine.DecodePath.decode(DecodePath.java:46)
12-25 14:24:24.457  5798  7575 E GlideExecutor: 	at com.bumptech.glide.load.engine.LoadPath.loadWithExceptionList(LoadPath.java:58)
12-25 14:24:24.457  5798  7575 E GlideExecutor: 	at com.bumptech.glide.load.engine.LoadPath.load(LoadPath.java:44)
12-25 14:24:24.457  5798  7575 E GlideExecutor: 	at com.bumptech.glide.load.engine.DecodeJob.runLoadPath(DecodeJob.java:437)
12-25 14:24:24.457  5798  7575 E GlideExecutor: 	at com.bumptech.glide.load.engine.DecodeJob.decodeFromFetcher(DecodeJob.java:430)
12-25 14:24:24.457  5798  7575 E GlideExecutor: 	at com.bumptech.glide.load.engine.DecodeJob.decodeFromData(DecodeJob.java:416)
12-25 14:24:24.457  5798  7575 E GlideExecutor: 	at com.bumptech.glide.load.engine.DecodeJob.decodeFromRetrievedData(DecodeJob.java:374)
12-25 14:24:24.457  5798  7575 E GlideExecutor: 	at com.bumptech.glide.load.engine.DecodeJob.onDataFetcherReady(DecodeJob.java:347)
12-25 14:24:24.457  5798  7575 E GlideExecutor: 	at com.bumptech.glide.load.engine.DataCacheGenerator.onDataReady(DataCacheGenerator.java:92)
12-25 14:24:24.457  5798  7575 E GlideExecutor: 	at com.bumptech.glide.load.model.ByteBufferFileLoader$ByteBufferFetcher.loadData(ByteBufferFileLoader.java:72)
12-25 14:24:24.457  5798  7575 E GlideExecutor: 	at com.bumptech.glide.load.engine.DataCacheGenerator.startNext(DataCacheGenerator.java:72)
12-25 14:24:24.457  5798  7575 E GlideExecutor: 	at com.bumptech.glide.load.engine.DecodeJob.runGenerators(DecodeJob.java:273)
12-25 14:24:24.457  5798  7575 E GlideExecutor: 	at com.bumptech.glide.load.engine.DecodeJob.runWrapped(DecodeJob.java:240)
12-25 14:24:24.457  5798  7575 E GlideExecutor: 	at com.bumptech.glide.load.engine.DecodeJob.run(DecodeJob.java:218)
12-25 14:24:24.457  5798  7575 E GlideExecutor: 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1113)
12-25 14:24:24.457  5798  7575 E GlideExecutor: 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:588)
12-25 14:24:24.457  5798  7575 E GlideExecutor: 	at java.lang.Thread.run(Thread.java:818)
12-25 14:24:24.457  5798  7575 E GlideExecutor: 	at com.bumptech.glide.load.engine.executor.GlideExecutor$DefaultThreadFactory$1.run(GlideExecutor.java:263)
12-25 14:24:24.462  5798  5798 E AndroidRuntime: FATAL EXCEPTION: main
```

**Comment by TWiStErRob** [comment]
Can you please edit the code above? There are some missing pieces:
 * the braces don't match up multiple times
 * `requestionOptions` (requestOptions?) is missing, it may contain leading evidence
 * in `onDestroy` you're iterating on the root views of `gallery_item_v2`. Glide won't clear an ImageView nested in it, only if the item is a single ImageView without a layout. Anyway onDestroy won't affect scrolling.

Does `context` refer to `ArrayAdapter.super.context`? Is that the activity or the app context? (I.e. does super constructor pass it on, or gets app context?

**Comment by TWiStErRob** [comment]
In the meantime I thought more about this: I think your stickers may be simply too big. Glide loads the frames of each sticker in the grid async... which means there must be space for 2 of the same sized Bitmaps: previous/current and next frame. This alone may already double your expectations, but possibly not the culprit. `GlideDrawable` holds the whole source GIF file in memory in a `byte[]`, with some parts decoded (e.g. frame offsets and metadata). So if the sticker GIFs are not well-compressed, they could be huge: for example they don't use the delta-encoding based on previous frame.

Here's a workaround I can think of: display a grid of stickers with `.asBitmap().diskCacheStrategy(ALL)` (~first frame) and let the user explicitly play the animation, but only one animation should play at any time. This means that you need some interaction (play button, long tap, selection, hover) to initiate playback which uses `.asGif().diskCacheStrategy(SOURCE)` to load the same file/uri into the same ImageView. (This is assumes remote stickers, if you have them on SD card already change: `ALL->RESULT` and `SOURCE->NONE`.)

**Comment by jcelaviie** [comment]
Thanks Robert for your quick response!
1. The requestOptions is : 
RequestOptions.overrideOf(360).centerInside(context);
2. in onDestroy I am actually iterating through stickerGridView and each item is AppCompatImageView:
android.support.v7.widget.AppCompatImageView{8195aa4 V.ED..... ......ID 13,13-223,223 #7f0d01fa app:id/gallery_item_image}
3. the arrayAdapter context refer to the activity.

In your suggested workaround, you recommend to change 
Glide.with(context)
          .asDrawable()
          .apply(requestionOptions)
          .load(item.photoUrl)
          .into(imageView);

to something like 
Glide.with(context)
          .asBitmap()
          .diskCacheStrategy(ALL)
          .apply(requestionOptions)
          .load(item.photoUrl)
          .into(imageView);

Thanks a lot!

**Comment by TWiStErRob** [comment]
Hmm, OK, so then everything should be fine based on override/onDestroy/context. Yes, that's the change, did it help?
I also assumed that you have animated Gif stickers, was that correct? I just saw you didn't mention anything like that, but the stack had a gift transformation.

Note that if you use the fragment as in `Glide.with(MyFragment.this)` you'll get the same behaviour you defined in onDestroy for free, see http://stackoverflow.com/a/32887693/253468

Please be aware that there's an open bug that prevents some memory to be freed, see #690.

**Comment by jcelaviie** [comment]
It doesn't help too much, the memory usage is still very high and crash while I load the last gridview fragment of GIF images (~ 50). It also will hit OOM if I scroll back and forth between these frames.

**Comment by jcelaviie** [comment]
Here are some more stack trace:
 art     : Clamp target GC heap from 207MB to 192MB
12-26 14:19:37.348 16292 17459 I art     : Alloc concurrent mark sweep GC freed 10(2400B) AllocSpace objects, 0(0B) LOS objects, 0% free, 191MB/192MB, paused 1.454ms total 67.980ms
12-26 14:19:37.353 16292 17459 W art     : Throwing OutOfMemoryError "Failed to allocate a 358572 byte allocation with 251232 free bytes and 245KB until OOM"
12-26 14:19:37.355 16292 17459 E GlideExecutor: Request threw uncaught throwable
12-26 14:19:37.355 16292 17459 E GlideExecutor: java.lang.OutOfMemoryError: Failed to allocate a 358572 byte allocation with 251232 free bytes and 245KB until OOM
12-26 14:19:37.355 16292 17459 E GlideExecutor: 	at dalvik.system.VMRuntime.newNonMovableArray(Native Method)
12-26 14:19:37.355 16292 17459 E GlideExecutor: 	at android.graphics.Bitmap.nativeCreate(Native Method)
12-26 14:19:37.355 16292 17459 E GlideExecutor: 	at android.graphics.Bitmap.createBitmap(Bitmap.java:831)
12-26 14:19:37.355 16292 17459 E GlideExecutor: 	at android.graphics.Bitmap.createBitmap(Bitmap.java:808)
12-26 14:19:37.355 16292 17459 E GlideExecutor: 	at android.graphics.Bitmap.createBitmap(Bitmap.java:775)
12-26 14:19:37.355 16292 17459 E GlideExecutor: 	at com.bumptech.glide.load.engine.bitmap_recycle.LruBitmapPool.getDirty(LruBitmapPool.java:140)
12-26 14:19:37.355 16292 17459 E GlideExecutor: 	at com.bumptech.glide.load.resource.bitmap.Downsampler.setInBitmap(Downsampler.java:475)
12-26 14:19:37.355 16292 17459 E GlideExecutor: 	at com.bumptech.glide.load.resource.bitmap.Downsampler.downsampleWithSize(Downsampler.java:317)
12-26 14:19:37.355 16292 17459 E GlideExecutor: 	at com.bumptech.glide.load.resource.bitmap.Downsampler.decodeFromWrappedStreams(Downsampler.java:174)
12-26 14:19:37.355 16292 17459 E GlideExecutor: 	at com.bumptech.glide.load.resource.bitmap.Downsampler.decode(Downsampler.java:145)
12-26 14:19:37.355 16292 17459 E GlideExecutor: 	at com.bumptech.glide.load.resource.bitmap.Downsampler.decode(Downsampler.java:104)
12-26 14:19:37.355 16292 17459 E GlideExecutor: 	at com.bumptech.glide.load.resource.bitmap.ByteBufferBitmapDecoder.decode(ByteBufferBitmapDecoder.java:33)
12-26 14:19:37.355 16292 17459 E GlideExecutor: 	at com.bumptech.glide.load.resource.bitmap.ByteBufferBitmapDecoder.decode(ByteBufferBitmapDecoder.java:17)
12-26 14:19:37.355 16292 17459 E GlideExecutor: 	at com.bumptech.glide.load.engine.DecodePath.decodeResourceWithList(DecodePath.java:69)
12-26 14:19:37.355 16292 17459 E GlideExecutor: 	at com.bumptech.glide.load.engine.DecodePath.decodeResource(DecodePath.java:54)
12-26 14:19:37.355 16292 17459 E GlideExecutor: 	at com.bumptech.glide.load.engine.DecodePath.decode(DecodePath.java:45)
12-26 14:19:37.355 16292 17459 E GlideExecutor: 	at com.bumptech.glide.load.engine.LoadPath.loadWithExceptionList(LoadPath.java:58)
12-26 14:19:37.355 16292 17459 E GlideExecutor: 	at com.bumptech.glide.load.engine.LoadPath.load(LoadPath.java:44)
12-26 14:19:37.355 16292 17459 E GlideExecutor: 	at com.bumptech.glide.load.engine.DecodeJob.runLoadPath(DecodeJob.java:437)
12-26 14:19:37.355 16292 17459 E GlideExecutor: 	at com.bumptech.glide.load.engine.DecodeJob.decodeFromFetcher(DecodeJob.java:430)
12-26 14:19:37.355 16292 17459 E GlideExecutor: 	at com.bumptech.glide.load.engine.DecodeJob.decodeFromData(DecodeJob.java:416)
12-26 14:19:37.355 16292 17459 E GlideExecutor: 	at com.bumptech.glide.load.engine.DecodeJob.decodeFromRetrievedData(DecodeJob.java:374)
12-26 14:19:37.355 16292 17459 E GlideExecutor: 	at com.bumptech.glide.load.engine.DecodeJob.onDataFetcherReady(DecodeJob.java:347)
12-26 14:19:37.355 16292 17459 E GlideExecutor: 	at com.bumptech.glide.load.engine.ResourceCacheGenerator.onDataReady(ResourceCacheGenerator.java:101)
12-26 14:19:37.355 16292 17459 E GlideExecutor: 	at com.bumptech.glide.load.model.ByteBufferFileLoader$ByteBufferFetcher.loadData(ByteBufferFileLoader.java:72)
12-26 14:19:37.355 16292 17459 E GlideExecutor: 	at com.bumptech.glide.load.engine.ResourceCacheGenerator.startNext(ResourceCacheGenerator.java:80)
12-26 14:19:37.355 16292 17459 E GlideExecutor: 	at com.bumptech.glide.load.engine.DecodeJob.runGenerators(DecodeJob.java:273)
12-26 14:19:37.355 16292 17459 E GlideExecutor: 	at com.bumptech.glide.load.engine.DecodeJob.runWrapped(DecodeJob.java:240)
12-26 14:19:37.355 16292 17459 E GlideExecutor: 	at com.bumptech.glide.load.engine.DecodeJob.run(DecodeJob.java:218)
12-26 14:19:37.355 16292 17459 E GlideExecutor: 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1113)
12-26 14:19:37.355 16292 17459 E GlideExecutor: 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:588)
12-26 14:19:37.355 16292 17459 E GlideExecutor: 	at java.lang.Thread.run(Thread.java:818)
12-26 14:19:37.355 16292 17459 E GlideExecutor: 	at com.bumptech.glide.load.engine.executor.GlideExecutor$DefaultThreadFactory$1.run(GlideExecutor.java:263)
12-26 14:19:37.358 16292 17466 I art     : Starting a blocking GC Alloc
12-26 14:19:37.358 16292 17466 I art     : Starting a blocking GC Alloc
12-26 14:19:37.368 16292 17466 I art     : Alloc sticky concurrent mark sweep GC freed 334(77KB) AllocSpace objects, 0(0B) LOS objects, 0% free, 191MB/192MB, paused 1.309ms total 9.510ms
12-26 14:19:37.373 16292 17466 I art     : Starting a blocking GC Alloc

**Comment by TWiStErRob** [comment]
Whoa-whoa-whoa, wait a minute... how many images are visible at once? And what do you mean when you say "last gridview fragment of GIF images": are there multiple instances of `MyFragment` added at the same time? Can you please post a little video or some images of the app? Or of at least a semantic image / mockup of the UI.

If there are 50 images visible at once (like a smiley selector) let's do some math: you used `override` of 360px, assuming square images:
```
360px*360px*4 bytes-per-pixel * 50 images = 506 kiB * 50 images = 24.7 MiB
```
There are devices with 16MB/24MB memory limits, even on a 32MB device you are likely to get OOM if my assumptions are correct. Did you try without using `override`? That will trigger Glide to use the layout size of the `ImageView` (make sure you're using match-parent or fixed dp sizes).

**Comment by TWiStErRob** [comment]
Also stack traces of `OutOfMemoryError`s don't help at all for diagnosing them. It just tells you it's broken and something filled up the memory. This filling happend much before the actual exception was thrown. In your case the only helpful information is that you have a memory class of 192MB. Try to get a heap dump and analize your usage, see https://github.com/bumptech/glide/issues/679#issuecomment-148194654

**Comment by jcelaviie** [comment]
There are 15 to 20 images are visible at once depending on the device rotation. There are 3 fragment instances created while first loading, and if I scroll back and forth, there will be one more fragment instance added each time, so the memory keep growing with loading more and more images. That is why I am trying to destroy the old fragment when I scroll to the new one. But it seems not working.
I have attached one picture of loading some images in one frame. Please take a look, thanks!
![device-2015-12-26-150554](https://cloud.githubusercontent.com/assets/16439842/12008245/73b1cc60-abe3-11e5-9fc3-a7caf8385899.png)

**Comment by TWiStErRob** [comment]
So we know that 192MB is available on your device, and we know that about one page of stickers needs 506 kiB * 20 ~ 10 MiB memory per page. Assuming this is Hangouts (it sure looks like it ;) there are 17 pages, so it's 170MiB loaded at once if you load everything on startup, which leaves 22MB left for the rest of the app. I just opened it and it uses 75MB as a baseline (open app, open most recent conversation with text only) so that's too much. This is why you want to load the pages dynamically. I had an assumption that these are all animated GIFs, but it looks like only the last rainbow group moves a little, so `asDrawable()`/`asBitmap()` shouldn't make much difference if these turles for example are PNG files.

Actually one-frame static GIFs were loaded as Bitmap instead of GifDrawable in v3 so that was better, but I couldn't find the same trick in v4 sources by a quick glance; it looks like for v4 the original GIF bytes and the first frame's Bitmap are both allocated while the image is displayed. @sjudd was this intentional, or I missed something?

**Does `onDestroy` get called when you're "trying to destroy the old fragment"?** I think that clearing code should be in `onDestroyView` since that may get called more often (e.g. when the view would get destroyed upon pushing to backstack) and it would be the symmetric counterpart of `onCreateView` where you set the adapter up.

So to sum up I'd just double-check which lifecycle events get called when scrolling horizontally and maybe reduce the  `offScreenPageLimit` of the `ViewPager` (if this is a `ViewPager`) and then if everything is fine, but the memory error still occurs take a dump and check who leaks those Bitmaps/other big chunks of memory that should be freed.

**Comment by sjudd** [comment]
There's nothing much for me to do here. As we've discussed, you're attempting to show too many and/or too large images. 

I'll file a separate issue for the single frame GIF enhancement, it doesn't appear to be the issue here.

**Comment by sjudd** [comment]
This is fixed.
